src/go/util/httppattern/uri_template_impl.go

```diff
@@ -0,0 +1,437 @@
+package httppattern
+
+import (
+	"bytes"
+	"fmt"
+)
+
+// The info about a variable binding {variable=subpath} in the template.
+type variable struct {
+	// Specifies the range of segments [start_segment, end_segment) the
+	// variable binds to. Both start_segment and end_segment are 0 based.
+	// end_segment can also be negative, which means that the position is
+	// specified relative to the end such that -1 corresponds to the end
+	// of the path.
+	StartSegment int
+	EndSegment   int
+
+	// The path of the protobuf field the variable binds to.
+	FieldPath []string
+
+	// Do we have a ** in the variable template?
+	HasDoubleWildCard bool
+}
+
+// Uri Template Grammar:
+// Questions:
+//
+// Template = "/" | "/" Segments [ Verb ] ;
+// Segments = Segment { "/" Segment } ;
+// Segment  = "*" | "**" | LITERAL | Variable ;
+// Variable = "{" FieldPath [ "=" Segments ] "}" ;
+// FieldPath = IDENT { "." IDENT } ;
+// Verb     = ":" LITERAL ;
+type parser struct {
+	input      string
+	tb         int
+	te         int
+	inVariable bool
+	segments   []string
+	verb       string
+	variables  []*variable
+}
+
+// `serializeUriTemplate` serialize the UriTemplate object into the string representation.
+// It assumes the input is valid uriTemplate generated from `Parse()` and it
+// won't do any validation.
+func serializeUriTemplate(input *UriTemplate, fieldNameMapping map[string]string) string {
```

Comment:
This is not just serialize?  It is serialize + replacement.   I am wondering if it is possible 
1) doing the replacement in UrlTemplate
2) then serailize() just do a searialize.

I think the code will be cleaner

Reply:
Done.